### PR TITLE
feat(surveys): Allow user to edit survey response adaptive limits

### PR DESCRIPTION
**Pull request by Phanatic**
## Problem

Allow user to edit survey response adaptive limits

## Changes

image

image

### CODE REVIEW

File: frontend/src/scenes/surveys/SurveyEdit.tsx [hunk 1]
```diff
@@ -79,11 +88,31 @@ export default function SurveyEdit(): JSX.Element {
...
+    useMemo(() => {
+        if (survey.responses_limit && survey.responses_limit > 0) {
+            setDataCollectionType('until_limit')
+        } else if (
+            survey.response_sampling_interval &&
+            survey.response_sampling_interval > 0 &&
+            survey.response_sampling_interval_type !== '' &&
+            survey.response_sampling_limit &&
+            survey.response_sampling_limit > 0
+        ) {
```
Comment (danielbachhuber) [inline frontend/src/scenes/surveys/SurveyEdit.tsx:110]: This logic isn't intuitive to me at first glance. Would it be worth adding a comment describing what it's supposed to do?

File: frontend/src/scenes/surveys/SurveyEdit.tsx [hunk 2]
```diff
@@ -852,44 +881,157 @@ export default function SurveyEdit(): JSX.Element {
...
+                                                    } else {
+                                                        resetSurveyResponseLimits()
+                                                        resetSurveyAdaptiveSampling()
+                                                    }
+                                                    setDataCollectionType(newValue)
+                                                }}
+                                                options={[
+                                                    {
+                                                        value: 'until_stopped',
+                                                        label: 'Keep collecting until stopped',
```
Comment (danielbachhuber) [inline frontend/src/scenes/surveys/SurveyEdit.tsx:922]:
```suggestion
                                                        label: 'Keep collecting responses until the survey is stopped',
```

File: frontend/src/scenes/surveys/SurveyEdit.tsx [hunk 3]
```diff
@@ -852,44 +881,157 @@ export default function SurveyEdit(): JSX.Element {
...
-                                                    responses are received.
-                                                    <Tooltip title="This is a rough guideline, not an absolute one, so the survey might receive slightly more responses than the limit specifies.">
-                                                        <IconInfo />
-                                                    </Tooltip>
-                                                </div>
-                                            )
-                                        }}
-                                    </LemonField>
+                                    <div className="mt-2">
+                                        <h3> How should we manage survey responses? </h3>
```
Comment (danielbachhuber) [inline frontend/src/scenes/surveys/SurveyEdit.tsx:885]:
```suggestion
                                        <h3>For how long would you like to receive survey responses?</h3>
```

File: frontend/src/scenes/surveys/SurveyEdit.tsx [hunk 4]
```diff
@@ -852,44 +881,157 @@ export default function SurveyEdit(): JSX.Element {
...
+                                                }}
+                                                options={[
+                                                    {
+                                                        value: 'until_stopped',
+                                                        label: 'Keep collecting until stopped',
+                                                        'data-attr': 'survey-collection-until-stopped',
+                                                    },
+                                                    {
+                                                        value: 'until_limit',
+                                                        label: 'Stop after reaching a limit',
```
Comment (danielbachhuber) [inline frontend/src/scenes/surveys/SurveyEdit.tsx:927]:
```suggestion
                                                        label: 'Stop displaying the survey after reaching a certain number of completed surveys',
```

File: frontend/src/scenes/surveys/SurveyEdit.tsx [hunk 5]
```diff
@@ -852,44 +881,157 @@ export default function SurveyEdit(): JSX.Element {
...
+                                                        'data-attr': 'survey-collection-until-limit',
+                                                    },
+                                                    {
+                                                        value: 'until_adaptive_limit',
+                                                        label: 'Adaptive limit by responses and time',
```
Comment (danielbachhuber) [inline frontend/src/scenes/surveys/SurveyEdit.tsx:932]:
```suggestion
                                                        label: 'Collect a certain number of surveys per day, week or month',
```

File: frontend/src/scenes/surveys/SurveyEdit.tsx [hunk 6]
```diff
@@ -852,44 +881,157 @@ export default function SurveyEdit(): JSX.Element {
...
+                                                        disabledReason: surveysAdaptiveLimitsDisabledReason,
+                                                    },
+                                                ]}
+                                            />
+                                        </LemonField.Pure>
+                                    </div>
+                                    {dataCollectionType == 'until_adaptive_limit' && (
+                                        <LemonField.Pure className="mt-4">
+                                            <div className="flex flex-row gap-2 items-center ml-5">
+                                                Starting{' '}
```
Comment (danielbachhuber) [inline frontend/src/scenes/surveys/SurveyEdit.tsx:943]:
```suggestion
                                                Starting on{' '}
```

File: frontend/src/scenes/surveys/SurveyEdit.tsx [hunk 7]
```diff
@@ -852,44 +881,157 @@ export default function SurveyEdit(): JSX.Element {
...
+                                                        />
+                                                    }
+                                                    visible={visible}
+                                                    onClickOutside={() => setVisible(false)}
+                                                >
+                                                    <LemonButton type="secondary" onClick={() => setVisible(!visible)}>
+                                                        {formatDate(dayjs(survey.response_sampling_start_date || ''))}
+                                                    </LemonButton>
+                                                </Popover>
+                                                , limit
```
Comment (danielbachhuber) [inline frontend/src/scenes/surveys/SurveyEdit.tsx:964]:
```suggestion
                                                , capture up to
```

File: frontend/src/scenes/surveys/SurveyView.tsx [hunk 8]
```diff
@@ -351,6 +351,21 @@ export function SurveyView({ id }: { id: string }): JSX.Element {
...
                                             )}
+                                            {survey.response_sampling_limit &&
+                                                survey.response_sampling_interval_type &&
+                                                survey.response_sampling_interval && (
+                                                    <>
+                                                        <span className="card-secondary mt-4">
+                                                            Completion conditions
+                                                        </span>
+                                                        <span>
+                                                            Survey response collection will be limited to receive{' '}
```
Comment (danielbachhuber) [inline frontend/src/scenes/surveys/SurveyView.tsx:362]:
```suggestion
                                                            Survey response collection is limited to receive{' '}
```

**Review by danielbachhuber — APPROVED** [review]
Looks good!

I did a full read-through of the code and left some comments with suggested text changes. I also tested the "edit survey" form and verified the various fields saved as expected.

Out of curiosity, were you able to test the limiting logic on the frontend, or are you simply relying on the tests? If the former, how did you test it?

File: frontend/src/scenes/surveys/SurveyEdit.tsx (hunk 1 above)
Comment (Phanatic) [reply frontend/src/scenes/surveys/SurveyEdit.tsx:110]: Good idea, doing that now!

**Pull request merged**